app/controllers/financialOverview/PaymentsAndChargesController.scala

```diff
@@ -0,0 +1,113 @@
+          case Some(fb) => aftConnector
+            .getAFTDetailsWithFbNumber(pstr, fb)
+            .map { aftDetails =>
+              val ua = UserAnswers(aftDetails.as[JsObject])
+              Seq(Tuple2(scheme.chargeType, (ua.get(AFTVersionQuery), ua.get(AFTReceiptDateQuery))))
+            }
+          case None => Future.successful(Nil)
+        }
+      }
+    }.map(_.flatten)
+    tuple.map(_.toMap)
+  }
+
+    // scalastyle:off method.length
+  def onPageLoad(srn: String, pstr: String, journeyType: ChargeDetailsFilter): Action[AnyContent] =
+    (identify andThen allowAccess()).async { implicit request =>
+      paymentsAndChargesService.getPaymentsForJourney(request.idOrException, srn, journeyType).flatMap { paymentsCache =>
+        val overdueCharges: Seq[SchemeFS] = paymentsAndChargesService.getOverdueCharges(paymentsCache.schemeFS)
+        val totalOverdue: BigDecimal = overdueCharges.map(_.amountDue).sum
+        val totalInterestAccruing: BigDecimal = overdueCharges.map(_.accruedInterestTotal).sum
+        val upcomingCharges: Seq[SchemeFS] = paymentsAndChargesService.extractUpcomingCharges(paymentsCache.schemeFS)
+        val totalUpcoming : BigDecimal = upcomingCharges.map(_.amountDue).sum
+
+        if (paymentsCache.schemeFS.nonEmpty) {
+          getMapChargeTypeToVersionAndDate(paymentsCache.schemeFS, pstr).flatMap { mapChargeTypesVersionsAndDate =>
+            val table = paymentsAndChargesService.getPaymentsAndCharges(srn, pstr, paymentsCache.schemeFS, mapChargeTypesVersionsAndDate, journeyType)
+            val tableOfPaymentsAndCharges = if (journeyType == Upcoming) removePaymentStatusColumn(table) else table
+            val json = Json.obj(
+              fields =
+                "titleMessage" -> Message(s"financialPaymentsAndCharges.$journeyType.title"),
+                "reflectChargeText" -> Message(s"financialPaymentsAndCharges.$journeyType.reflect.charge.text"),
+                "journeyType" -> journeyType.toString,
+                "paymentAndChargesTable" -> tableOfPaymentsAndCharges,
+                "schemeName" -> paymentsCache.schemeDetails.schemeName,
+                "totalOverdue" -> s"${FormatHelper.formatCurrencyAmountAsString(totalOverdue)}",
+                "totalInterestAccruing" -> s"${FormatHelper.formatCurrencyAmountAsString(totalInterestAccruing)}",
+                "totalUpcoming" -> s"${FormatHelper.formatCurrencyAmountAsString(totalUpcoming)}",
+                "returnUrl" -> config.schemeDashboardUrl(request).format(srn)
+            )
+            renderer.render(template = "financialOverview/paymentsAndCharges.njk", json).map(Ok(_))
+          }
+        } else {
+          Future.successful(Redirect(controllers.routes.SessionExpiredController.onPageLoad))
+        }
+      }
+    }
+
+  private val removePaymentStatusColumn: Table => Table = table =>
+    Table(table.caption, table.captionClasses, table.firstCellIsHeader,
+      table.head.take(table.head.size - 1),
```

Comment:
I'd use headOption here just in case the table is empty

Reply:
Not available to use as option